### PR TITLE
Always Pre-Split Microbatches for PP

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3856

Summary:
- Have PP training and validation dataloaders emit pipeline microbatches directly.
- Keep PP execution in the existing forward_backward_step/eval flow while passing pre-split args/kwargs/targets to torch.distributed.pipelining.
- Remove the PP+Varlen guard and switch the GPT-OSS PP integration test back to the default varlen config.
- Update copied PP train paths in TorchFT and Forge to use the pre-split schedule API with the same structure-preserving pattern.

Motivation:
TorchTitan previously built a full local batch and relied on torch.distributed.pipelining to split it. That does not work for batch-dependent non-tensor metadata such as varlen attention metadata. Owning the split at the dataloader/trainer boundary lets TorchTitan generate block masks and varlen metadata per pipeline microbatch before calling the PP schedule.

Test Plan:
- python -m py_compile torchtitan/trainer.py torchtitan/components/validate.py torchtitan/models/common/decoder.py torchtitan/models/gpt_oss/config_registry.py tests/integration_tests/models.py torchtitan/experiments/torchft/trainer.py torchtitan/experiments/forge/engine.py torchtitan/experiments/forge/example_train.py
- pre-commit run --files torchtitan/trainer.py torchtitan/components/validate.py torchtitan/models/common/decoder.py torchtitan/models/gpt_oss/config_registry.py tests/integration_tests/models.py torchtitan/experiments/torchft/trainer.py torchtitan/experiments/forge/engine.py torchtitan/experiments/forge/example_train.py
- python -m tests.integration_tests.run_tests <output_dir> --test_suite models --test_name gpt_oss_pp+fsdp+ep+sacop --ngpu 8
- NGPU=2 LOG_RANK=0,1 ./run_train.sh --dump_folder <output_dir> --module gpt_oss --config gpt_oss_debugmodel --parallelism.pipeline_parallel_degree 2 --parallelism.pipeline_parallel_schedule 1F1B --training.steps 1 --validator.enable --validator.steps 1

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -740,6 +767,10 @@ def forward_backward_step(
             else:
                 loss = torch.tensor([-1.0], device=self.device)
```
Comment (fegin) [inline torchtitan/trainer.py]: Let's move this to a method `pp_forward_backward_step` and this block becomes:

```
if parallel_dims.pp_enabled:
     return self.pp_forward_backward_step(...)

non_pp_logic
```

Reason: the if/else blocks now share nothing except for local alias. Before this PR, the two at least share the inputs, labels, extra_kwargs = self.post_dataloading_process(input_dict, labels). And the pp logic becomes more complicated now and justifies a method for readability.

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -788,15 +835,26 @@ def train_step(
...
+                            mb_input_dict[k] = v.to(self.device)
+                labels = [mb_labels.to(self.device) for mb_labels in labels]
+            else:
+                assert isinstance(input_dict, dict)
+                assert isinstance(labels, torch.Tensor)
+                # Move tensors to GPU
+                for k, v in input_dict.items():
+                    if isinstance(v, torch.Tensor):
+                        input_dict[k] = v.to(self.device)
+                labels = labels.to(self.device)
```
Comment (fegin) [inline torchtitan/trainer.py]: We cannot use PP with gradient accumulation after this PR, right? If so, do we error out somewhere if users try to use them together?

File: torchtitan/trainer.py [hunk 3]
```diff
@@ -788,15 +835,26 @@ def train_step(
...
 
             loss = self.forward_backward_step(
                 input_dict=input_dict,
-                labels=labels,
+                labels=labels,  # pyrefly: ignore [bad-argument-type]
```
Comment (fegin) [inline torchtitan/trainer.py]: Why do we need `pyrefly ignore` now? Let's try to avoid using `pyrefly ignore`. Let's use cast if we really cannot figure out why.

File: torchtitan/trainer.py (hunk 2 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Using pp with gradient accumulation doesn't make sense. PP is already doing gradient accumulation. So we can disable gradient accumulation for pp.

Comment (tianyu-l) [reply torchtitan/trainer.py]: I don't think that's true. See previous discussions in https://github.com/pytorch/torchtitan/issues/1733

Conclusion is that we should continue to support gradient accumulation + PP.

File: torchtitan/trainer.py [hunk 4]
```diff
@@ -764,15 +795,31 @@ def train_step(
         # Keep these variables local to shorten the code as these are
         # the major variables that are used in the training loop.
         parallel_dims = self.parallel_dims
+        num_microbatches = (
+            self.config.training.local_batch_size
+            // self.config.parallelism.pipeline_parallel_microbatch_size
+            if parallel_dims.pp_enabled
+            else 1
+        )
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: This is computed twice in constructor and here

File: torchtitan/trainer.py [hunk 5]
```diff
@@ -764,15 +795,31 @@ def train_step(
...
-        microbatches = []
+        microbatches: list[tuple[Any, Any]] = []
         local_valid_tokens = torch.tensor(0, dtype=torch.int64)
-        for _microbatch in range(self.gradient_accumulation_steps):
-            with sl.log_trace_span("fetching_batch"):
-                input_dict, labels = next(data_iterator)
-                local_valid_tokens += (labels != IGNORE_INDEX).sum()
-                microbatches.append((input_dict, labels))
+        for _ in range(self.gradient_accumulation_steps):
+            if parallel_dims.pp_enabled:
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: it feels to me this can be improved -- two pieces of code seem almost identical

File: torchtitan/trainer.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/trainer.py]: The main change / challenge in this PR is to properly differentiate PP microbatching vs. gradient accumulation microbatching, from UX / config perspective.

E.g. a simple way is to let user specify "local microbatch size"
- without PP it's the `local_batch_size` today, which is used together with `global_batch_size` to determine number of grad accum steps
- with PP, it's the local microbatch size, whose product with num PP microbatches * num grad accum steps would have to equal `global_batch_size`.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

**Comment by acisseJZhong** [comment]
synced with @sanketpurandare, he will support both PP + grad accum in this PR first. After that I will remove batch dim, and @jinsooihm can also work on refactoring `post_dataloading_process`. 

cc @fegin @tianyu-l

File: torchtitan/trainer.py [hunk 6]
```diff
@@ -502,9 +509,15 @@ def __init__(self, config: Config):
             dp_rank=batch_rank,
             tokenizer=self.tokenizer,
             seq_len=config.training.seq_len,
-            local_batch_size=config.training.local_batch_size,
+            local_batch_size=(
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: seems confusing, maybe should rename dataloader.local_batch_size to microbatch_size? although we'd kill this arg soon anyway

File: torchtitan/trainer.py [hunk 7]
```diff
@@ -700,63 +713,77 @@ def post_dataloading_process(
...
         # The returned loss here is local SUM loss / global_valid_tokens
         return loss
 
+    def pp_forward_backward_step(
+        self,
+        *,
+        input_dict_mbs: list[dict[str, torch.Tensor]],
+        label_mbs: list[torch.Tensor],
+        global_valid_tokens: float,
+    ) -> torch.Tensor:
```
Comment (tianyu-l) [inline torchtitan/trainer.py:773]: I feel it's fine to not have this function, to avoid
- asymmetric code structure between PP and non-PP
- giving new names like `input_dict_mbs`

File: torchtitan/experiments/forge/example_train.py [hunk 8]
```diff
@@ -265,15 +279,15 @@ def train_step(
         # Keep these variables local to shorten the code as these are
         # the major variables that are used in the training loop.
         parallel_dims = self.parallel_dims
-
-        # Collect all microbatches on CPU and count total valid tokens
-        # Here we assume the inputs/labels are on GPU
-        microbatches = []
+        batch_groups: list[list[tuple[dict[str, torch.Tensor], torch.Tensor]]] = []
```
Comment (tianyu-l) [inline torchtitan/experiments/forge/example_train.py]: name too general, can we achieve clear semantics that
- batch_groups is containing the list of microbatches for each optimizer step
- each batch_group contains the microbatches for each fwd-bwd step

File: torchtitan/experiments/forge/example_train.py [hunk 9]
```diff
@@ -285,18 +299,28 @@ def train_step(
...
+            else:
+                assert len(input_dict_mbs) == len(label_mbs) == 1
+                step_input_dict = input_dict_mbs[0]
+                step_labels = label_mbs[0]
 
             loss = self.forward_backward_step(
-                input_dict=input_dict,
-                labels=labels,
+                input_dict=step_input_dict,
+                labels=step_labels,
```
Comment (tianyu-l) [inline torchtitan/experiments/forge/example_train.py]: above you called it batch_group, here it's step_X, need to be consistent

File: torchtitan/trainer.py (hunk 6 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Pulled the conditional into a `dataloader_batch_size` local before the build
call. I kept the shared `local_batch_size` constructor keyword unchanged because
all dataloader implementations use it and, as noted, it is going away; renaming
that contract here would create unrelated churn.

**Comment by sanketpurandare** [comment]
Updated the PR to support PP and gradient accumulation independently. Each
gradient-accumulation iteration now groups its pipeline microbatches before one
schedule invocation. A 2-GPU PP + gradient-accumulation run and a deterministic
10-step comparison both passed; loss and grad norm matched bitwise at every
step.

File: torchtitan/trainer.py (hunk 7 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py:773]: Kept the extracted method because it directly addressed @fegin 's earlier
request in
https://github.com/pytorch/torchtitan/pull/3856#discussion_r3515161429. The PP
branch owns list-typed preprocessing and schedule construction, while the non-PP
path remains linear. I did not inline it because that would reverse the earlier
requested structure.

File: torchtitan/experiments/forge/example_train.py (hunk 8 above)
Comment (sanketpurandare) [reply torchtitan/experiments/forge/example_train.py]: Renamed `batch_groups` to `microbatch_groups` and each `batch_group` to
`microbatches` across core, TorchFT, and Forge. The comment now states that all
groups form one optimizer step and each group feeds one forward-backward call.

File: torchtitan/experiments/forge/example_train.py (hunk 9 above)
Comment (sanketpurandare) [reply torchtitan/experiments/forge/example_train.py]: Renamed `step_input_dict` and `step_labels` to `fwd_bwd_input_dict` and
`fwd_bwd_labels` across all three paths, matching the unit represented by each
`microbatches` group.

File: torchtitan/trainer.py (hunk 1 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Extracted the PP path into `pp_forward_backward_step`. The dispatcher now
returns through that method when PP is enabled, leaving the non-PP path linear.

File: torchtitan/trainer.py (hunk 2 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Kept PP microbatching distinct from gradient accumulation. Each local batch is
split into `num_pipeline_parallel_microbatches`, and each gradient-accumulation
iteration invokes the PP schedule once with that group. The config docs now
state both batch-size relationships, and the combined path passed the 2-GPU
test plus a bitwise-identical 10-step loss and grad-norm comparison.

File: torchtitan/trainer.py (hunk 3 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Removed the `pyrefly` suppression. The dispatcher uses the explicit batch
union and narrows it with the PP/non-PP invariants; the pyrefly hook passes.

File: torchtitan/trainer.py (hunk 4 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Computed `num_pipeline_parallel_microbatches` once during trainer construction
and reused the cached value for dataloader checkpoint cadence and training.

File: torchtitan/trainer.py (hunk 5 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py]: Collapsed the duplicated PP and non-PP fetch branches into one loop. Each
gradient-accumulation iteration now builds one microbatch group whose size is
one without PP or `num_pipeline_parallel_microbatches` with PP.

**Review by tianyu-l — APPROVED** [review]
LGTM, thanks!

File: torchtitan/config/configs.py [hunk 10]
```diff
@@ -200,9 +204,7 @@ class ParallelismConfig:
     pipeline_parallel_microbatch_size: int = 1
     """
     The size of each pipeline parallel microbatch (default 1).
-    This value is used to compute the total number of microbatches by dividing local_batch_size with
-    pipeline_parallel_microbatch_size.
-    The global training batch size must be evenly divisible by pipeline_parallel_microbatch_size.
+    `training.local_batch_size` must be evenly divisible by this value.
```
Comment (acisseJZhong) [inline torchtitan/config/configs.py:207]: shall we assert this in post init?

File: torchtitan/trainer.py [hunk 11]
```diff
@@ -700,63 +714,77 @@ def post_dataloading_process(
...
-                    self.pp_schedule.step(
-                        **extra_kwargs,
-                        target=targets,
-                        losses=losses,
-                        loss_kwargs=loss_kwargs,
-                        return_outputs=False,
-                    )
+            assert isinstance(input_dict, list)
+            assert isinstance(labels, list)
+            return self.pp_forward_backward_step(
```
Comment (acisseJZhong) [inline torchtitan/trainer.py:744]: instead of having 
```
forward_backward_step()
- if pp_enabled: self.pp_forward_backward_step()
- else: normal forward backward step
```
can we move it outside to `train_step`? 
```
in train_step():
- if pp_enabled: self.pp_forward_backward_step()
- else: self.forward_backward_step()
```
would this makes things clearer?

Comment (tianyu-l) [reply torchtitan/trainer.py:744]: I thought about that. My counter argument would be
- if some repo is integrating torchtitan Trainer, e.g. in their RL, it's possible that they'd prefer calling a single `trainer.forward_backward_step()` than conditioning on PP

but maybe that example is too made-up haha

Comment (acisseJZhong) [reply torchtitan/trainer.py:744]: valid enough, given that miles is integrating with titan, I think we need this API to work. 

we can always refactor this trainer API later if needed.

File: torchtitan/trainer.py [hunk 12]
```diff
@@ -788,18 +818,29 @@ def train_step(
...
+                input_dict_mbs.append(input_dict)
+                label_mbs.append(labels.to(self.device))
+
+            if parallel_dims.pp_enabled:
+                fwd_bwd_input_dict = input_dict_mbs
+                fwd_bwd_labels = label_mbs
+            else:
+                assert len(input_dict_mbs) == len(label_mbs) == 1
+                fwd_bwd_input_dict = input_dict_mbs[0]
+                fwd_bwd_labels = label_mbs[0]
```
Comment (acisseJZhong) [inline torchtitan/trainer.py:856]: can we just pass in input_dict_mbs and label_mbs into  `forward_backward_step`? and assert length == 1 inside `forward_backward_step` when PP off?

File: torchtitan/config/configs.py (hunk 10 above)
Comment (sanketpurandare) [reply torchtitan/config/configs.py:207]: Added the cross-field check to `Trainer.Config.__post_init__`, where both config values are available after CLI overrides. It now raises `ValueError` for a non-positive pipeline microbatch size and for PP configurations where the local batch size is not evenly divisible. Added an end-to-end config parsing test for the divisibility check.

File: torchtitan/trainer.py (hunk 12 above)
Comment (sanketpurandare) [reply torchtitan/trainer.py:856]: Kept the one-element unwrap at the `train_step` boundary. Moving it into `forward_backward_step` would change the existing non-PP contract from a dict/tensor pair to lists; direct callers such as GraphTrainer use the current contract, and the adjacent thread established this method as an integration API. A list-only contract should be an intentional API migration rather than a callsite cleanup in this PR.
